backend/tests/test_treasury_cacher.py

```diff
@@ -0,0 +1,166 @@
+from json import JSONDecodeError, loads
+
+from fakeredis import FakeRedis
+from httpx import Client
+from pytest import MonkeyPatch, mark, raises
+
+from backend.app.libs.storage_helpers.storage_helpers import (
+    retrieve_treasuries_metadata,
+)
+from backend.app.libs.storage_helpers.treasury_cacher import (
+    _get_treasury_list,
+    get_and_store_treasury_list,
+)
+
+from .conftest import (
+    HTTPStatusError,
+    raise_http_status_error_404,
+    raise_http_status_error_501,
+)
+
+
+class MockResponse:
+    status_code = None
+
+    def raise_for_status(self):
+        return None
+
+    @staticmethod
+    def json():
+        return {
+            "success": True,
+            "data": [
+                {
+                    "id": "mock_ns",
+                    "bundle": None,
+                    "results": {"mocked_tres_portfolios"},
+                    "metadata": {
+                        "icon": "data:image/svg+xm;base64,mocked_b64_img",
+                        "name": "Mock_NS",
+                        "website": "mockns.domains.com",
+                        "treasuries": [
+                            "0x6d6f636b65645f74726561737572795f31",
+                            "0x6d6f636b65645f74726561737572795f32",
+                            "0x6d6f636b65645f74726561737572795f33",
+                        ],
+                    },
+                }
+            ],
+        }
+
+
+def test_get_treasury_list_success(monkeypatch: MonkeyPatch):
+    monkeypatch.setattr(Client, "get", lambda *_: MockResponse())
+
+    treasury_list = _get_treasury_list()
+    assert treasury_list
```

Comment:
Can the assertion rather describe what the whole `treasury_list` variable is equal to?

Like: `asset treasury_list == …` (fill in the blanks).

Reply:
The contents of `treasuries` is tested in this section

```python
    treasury_addresses = [
        "0x6d6f636b65645f74726561737572795f31",
        "0x6d6f636b65645f74726561737572795f32",
        "0x6d6f636b65645f74726561737572795f33",
    ]
    # ...
    for treasury in treasuries:
        assert treasury[1] == 1
        assert isinstance(treasury[0], str)
        if treasury[0] in treasury_addresses:
            all_included += 1
    assert all_included == 3
```

Which confirms that each element in `treasuries` contains a `str` and a number, further is defines the expected values for these `str`s and numbers.

I will add more specific type assertions in my next commit, for the `treasuries` data structure and the data structure of its elements (`list` and `tuple`s respectively) as well as for element 1 of each element of `treasuries` (`int`).

The problem with `assert treasury_list == …` is the ordering, e.g. `[1,3,4] == [1, 4, 3]` evaluates as false. This is relevant because elements in the `treasuries` key on redis are elements of a redis set, which is unordered. When the response for `smembers` is converted to a list in `store_treasury_metadata` the ordering is not necessarily the same as how they were `sadd`ed.

Reply:
E.g. of responses from `retrieve_treasuries_metadata`:

Call 1) 
![image](https://user-images.githubusercontent.com/13321394/176312152-cb2e8c19-3583-4e06-aabd-0fe09988ffff.png)

Call 2) 
![image](https://user-images.githubusercontent.com/13321394/176312209-884dbaf5-ec8c-407a-a937-c6186b318a1b.png)

Call 3) 
![image](https://user-images.githubusercontent.com/13321394/176312256-7df0a002-b276-4359-b024-46fab9965d70.png)

Reply:
> The problem with assert treasury_list == … is the ordering, e.g. [1,3,4] == [1, 4, 3] evaluates as false. This is relevant because elements in the treasuries key on redis are elements of a redis set, which is unordered. When the response for smembers is converted to a list in store_treasury_metadata the ordering is not necessarily the same as how they were sadded.

Yes and that's precisely why testing against the whole value is more valuable: it allows us to see, understand and master the underlying abstractions.

It seems to me you have 2 choices:

- Keep `retrieve_treasuries_metadata` as is, returning a list, and make a test that explicitely shows that the returned list is garbled.
- Change `retrieve_treasuries_metadata` so that it returns a set and test `assert retrieve_treasuries_metadata(something) == set(something)`.

⚠️ Your test should not do something like `set(retrieve_treasuries_metadata(something)) == set(something)`. This would be too weak an assertion.

Reply:
Can we please make an assertion like

```python
assert treasury_list == ['0xsomething', '0xsomething']
```
?